### PR TITLE
Implement query parameter management.

### PR DESCRIPTION
**Pull request by swankjesse**
The tricky part of this is '+' and ' ' management.

### CODE REVIEW

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java [hunk 1]
```diff
@@ -612,4 +613,70 @@
...
+    HttpUrl base = HttpUrl.parse("http://host/");
+    HttpUrl url = base.newBuilder().addQueryParameter("a+=& b", "c+=& d").build();
+    assertEquals("http://host/?a%2B%3D%26+b=c%2B%3D%26+d", url.toString());
+    assertEquals("c+=& d", url.queryParameterValue(0));
+    assertEquals("a+=& b", url.queryParameterName(0));
+    assertEquals("c+=& d", url.queryParameter("a+=& b"));
+    assertEquals(Collections.singleton("a+=& b"), url.queryParameterNames());
+    assertEquals(Collections.singletonList("c+=& d"), url.queryParameterValues("a+=& b"));
+    assertEquals(1, url.querySize());
+    assertEquals("a+=& b=c+=& d", url.decodeQuery()); // Ambiguous! (Though working as designed.)
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Does this method have a use case? Or is it just for parity?

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java [hunk 2]
```diff
@@ -612,4 +613,70 @@
...
+    assertEquals("c =& d", url.queryParameter("a =& b"));
+  }
+
+  @Test public void composeQueryRemoveQueryParameter() throws Exception {
+    HttpUrl url = HttpUrl.parse("http://host/").newBuilder()
+        .addQueryParameter("a+=& b", "c+=& d")
+        .removeAllQueryParameters("a+=& b")
+        .build();
+    assertEquals("http://host/", url.toString());
+    assertEquals(null, url.queryParameter("a+=& b"));
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: What does query size return for this case?

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java [hunk 3]
```diff
@@ -612,4 +613,70 @@
...
+        .removeAllEncodedQueryParameters("a+=& b")
+        .build();
+    assertEquals("http://host/", url.toString());
+    assertEquals(null, url.queryParameter("a =& b"));
+  }
+
+  @Test public void composeQuerySetQueryParameter() throws Exception {
+    HttpUrl url = HttpUrl.parse("http://host/").newBuilder()
+        .addQueryParameter("a+=& b", "c+=& d")
+        .setQueryParameter("a+=& b", "ef")
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Duplicate this test and do two calls to add and ensure repeated names are allowed.

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 2 above)
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: I guess it returns 0 from seeing the code. Should it return -1 when empty to differentiate between `http://host/` and `http://host/?` cases?

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 4]
```diff
@@ -362,37 +365,64 @@ public String decodeQuery() {
...
-    throw new UnsupportedOperationException(); // TODO(jwilson).
+    if (queryNamesAndValues == null) return null;
+    String encodedName = canonicalize(name, QUERY_COMPONENT_ENCODE_SET, false, true);
+    for (int i = 0, size = queryNamesAndValues.size(); i < size; i += 2) {
+      if (encodedName.equals(queryNamesAndValues.get(i))) {
+        String value = queryNamesAndValues.get(i + 1);
+        return value != null ? percentDecode(value, true) : null;
+      }
+    }
+    return null;
```
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: This is ambiguous between "not found" and "found but value was null".

When is the value `null`? `http://host/?foo`? `http://host/?foo=`? I would think the last one would be empty string. Not sure what the behavior of the former would be.

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 1 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: As far as I can tell, there's applications that don't want to cut the query up into name/value pairs, and instead want a single string. This is for those applications.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java [hunk 5]
```diff
@@ -569,48 +599,75 @@ public Builder encodedQuery(String encodedQuery) {
...
+      if (queryNamesAndValues == null) return this;
+      String nameToRemove = canonicalize(name, QUERY_COMPONENT_ENCODE_SET, false, true);
+      removeAllCanonicalQueryParameters(nameToRemove);
+      return this;
     }
 
     public Builder removeAllEncodedQueryParameters(String encodedName) {
-      if (encodedName == null) throw new IllegalArgumentException("encodedName == null");
-      throw new UnsupportedOperationException(); // TODO(jwilson)
+      if (encodedName == null) throw new IllegalArgumentException("name == null");
```
Comment (JakeWharton) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Wrong param name in exception.

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 1 above)
Comment (JakeWharton) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Wouldn't that be `query()` though?

**Comment by JakeWharton** [comment]
I'd be interested in how the following behaves:
 * `http://host/?foo&bar&baz`
 * `http://host/?foo=&bar=&baz=`
 * `http://host/?foo[]=1&foo[]=2&foo=[3]`

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 1 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: no, `query()` is still encoded. (Until I flip the defaults, at which point this becomes `query()` and that become `encodedQuery()`.)

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 2 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Fixed as discussed, plus tests.

File: okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java (hunk 3 above)
Comment (swankjesse) [inline okhttp-tests/src/test/java/com/squareup/okhttp/HttpUrlTest.java]: Done.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java (hunk 4 above)
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: If there's an equals sign, the value is non-null.

File: okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java (hunk 5 above)
Comment (swankjesse) [inline okhttp/src/main/java/com/squareup/okhttp/HttpUrl.java]: Good catch.

**Comment by swankjesse** [comment]
Please take a look! I still need to address `+` vs `%20` when encoding spaces.

**Comment by JakeWharton** [comment]
Also a rebase!

**Comment by swankjesse** [comment]
Rebased.

**Comment by JakeWharton** [comment]
LGTM. These tests are great.

**Comment by swankjesse** [comment]
https://github.com/square/okhttp/issues/1486

**Pull request merged**